### PR TITLE
feat(gitlab): retry requests on resource locks

### LINKED ISSUE #29006: Retry GitLab API requests on HTTP 409 Resource lock errors
### Describe the proposed change(s).

I'm creating this issue based on a comment from @fgreinacher in discussion #28972. I edited the comment for readability.

The GitLab API may return `HTTP 409 Resource lock` in cases where an internal DB lock fails to be acquired in time, see:

- https://gitlab.com/gitlab-org/gitlab/-/blob/443c12cf3b238385db728f03b2cdbb4f17c70292/lib/api/api.rb#L111

Those errors are usually transient and a retry is therefore worthwhile.

For example, the `python-gitlab` library added special handling for this in [python-gitlab/python-gitlab#2326](https://github.com/python-gitlab/python-gitlab/pull/2326).

We should add similar logic to Renovate's GitLab layer.

### PR DESCRIPTION
**Pull request by fgreinacher**
## Changes

This makes Renovate retry requests to the GitLab API when they fail with a resource lock error.

## Context

Closes https://github.com/renovatebot/renovate/issues/29006

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

:hammer_and_pick: with :heart: by [Siemens](https://opensource.siemens.com/)

### CODE REVIEW

File: lib/util/http/index.ts [hunk 1]
```diff
@@ -124,8 +129,26 @@ export class Http<Opts extends HttpOptions = HttpOptions> {
...
+            }
+            if (error.response && retryAfter) {
+              // Don't rely on `got` retry-after handling, just let it fail and then we'll handle it
+              return 0;
+            }
+            if (!this.shouldRetryOnError(error, retryOptions)) {
+              return 0;
+            }
+            const noise = Math.random() * 100;
+            return 2 ** (attemptCount - 1) * 1000 + noise;
```
Comment (fgreinacher) [inline lib/util/http/index.ts:149]: Taken from https://github.com/sindresorhus/got/blob/88e623a0d8140e02eef44d784f8d0327118548bc/source/core/calculate-retry-delay.ts#L5

File: lib/util/http/index.ts [hunk 2]
```diff
@@ -248,6 +271,24 @@ export class Http<Opts extends HttpOptions = HttpOptions> {
...
+    const hasStatusCode =
+      error.response &&
+      retryOptions.statusCodes.includes(error.response.statusCode);
+    if (!hasMethod || (!hasErrorCode && !hasStatusCode)) {
+      return false;
+    }
+    if (error.response && error.response.statusCode === 413) {
+      return false;
+    }
+    return true;
```
Comment (fgreinacher) [inline lib/util/http/index.ts:289]: Taken from https://github.com/sindresorhus/got/blob/88e623a0d8140e02eef44d784f8d0327118548bc/source/core/calculate-retry-delay.ts#L5.

File: lib/util/http/gitlab.ts [hunk 3]
```diff
@@ -83,4 +89,19 @@ export class GitlabHttp extends Http<GitlabHttpOptions> {
...
+  protected override calculateRetryDelay(retryObject: RetryObject): number {
+    const { error, attemptCount, retryOptions } = retryObject;
+    if (
+      attemptCount <= retryOptions.limit &&
+      error.options.method === 'POST' &&
+      error.response?.statusCode === 409 &&
+      error.response.rawBody.toString().includes('Resource lock')
+    ) {
+      const noise = Math.random() * 100;
+      return 2 ** (attemptCount - 1) * 1000 + noise;
```
Comment (fgreinacher) [inline lib/util/http/gitlab.ts:102]: Copied from https://github.com/sindresorhus/got/blob/v11.8.6/source/core/calculate-retry-delay.ts#L33-L34

File: lib/util/http/gitlab.ts [hunk 4]
```diff
@@ -83,4 +84,19 @@ export class GitlabHttp extends Http<GitlabHttpOptions> {
       throw err;
     }
   }
+
+  protected override calculateRetryDelay(retryObject: RetryObject): number {
+    const { error, attemptCount, retryOptions } = retryObject;
```
Comment (viceice) [inline lib/util/http/gitlab.ts:89]:
```suggestion
  protected override calculateRetryDelay({ error, attemptCount, retryOptions }: RetryObject): number {
```
why not use destruction in function args, likle you do in base class ?

Comment (fgreinacher) [reply lib/util/http/gitlab.ts:89]: Much better indeed, not sure why I haven't done it from the start 🤔

Comment (fgreinacher) [reply lib/util/http/gitlab.ts]: Ah, now I remember: We need the complete `retryObject` for passing it to the base class.

**Review by viceice — APPROVED** [review]

**Pull request merged**